src/c5t.js

```diff
@@ -0,0 +1,385 @@
+   * @returns {boolean}
+   */
+  function _isPrimitiveType(value) {
+    return (-1 < _primitiveTypes.indexOf(typeof value));
+  }
+  
+  /**
+   * Shallow extend of one object with another.
+   * Does not propagate non-primitive data types to the destination.
+   *
+   * @param {Object} dst The destination object.
+   * @param {Object} [src] The source object.
+   * @param {Function} [predicate] The predicate to check if the property should be copied.
+   * @returns {Object} The destination object.
+   */
+  function _extend(dst, src, predicate) {
+    if (src) {
+      for (var x in src) {
+        if (_hasOwnProperty[str_call](src, x) && (predicate ? predicate(src[x], x) : true)) {
+          dst[x] = src[x];
+        }
+      }
+    }
+    return dst;
+  }
+  
+  var _transports = {};
+  
+  _transports[str_xhr] = function (baseUrl, payloadString, callbackFn) {
+    var xhr = new XMLHttpRequest();
+    xhr.onreadystatechange = function () {
+      if (xhr.readyState === 4) {
+        xhr.onreadystatechange = null;
+        xhr = null;
+        callbackFn();
+      }
+    };
+    // Use synchronous request.
+    xhr.open('POST', baseUrl, false);
+    xhr.send(payloadString);
+  };
+  
+  _transports[str_image] = function (baseUrl, payloadString, callbackFn) {
+    var image = new Image();
+    image.onload = image.onerror = function () {
+      image.onload = image.onerror = null;
+      image = null;
+      callbackFn();
+    };
+    image.src = baseUrl + '?' + payloadString;
+  };
+  
+  // If the `sendBeacon` API is not supported we don't add the transport implementation.
+  if (navigator.sendBeacon) {
+    _transports[str_beacon] = function (baseUrl, payloadString, callbackFn) {
+      navigator.sendBeacon(baseUrl, payloadString);
+      setTimeout(callbackFn, 0);
+    };
+  }
+  
+  /**
+   * Serializes the hit info into application/x-www-form-urlencoded format.
+   *
+   * @param {Object} trackerState The current state of the tracker.
+   * @param {string} hitType The hit type, e.g. 'event' or 'pageview'.
+   * @param {Object} [fieldObject] The overrides for this particular hit.
+   * @returns {string} The serialized hit info.
+   */
+  function _serializeHit(trackerState, hitType, fieldObject) {
+    var payload = _extend(_extend({}, trackerState), fieldObject);
+    payload[str_hitType] = hitType;
+    delete payload[str_trackingId];
+    delete payload[str_useBeacon];
+    delete payload[str_transport];
+    var payloadString = '';
+    for (var x in payload) {
+      if (_hasOwnProperty[str_call](payload, x) && _isPrimitiveType(payload[x])) {
+        payloadString += (
+          (payloadString ? '&' : '') +
+          encodeURIComponent(x) + '=' +
+          encodeURIComponent(payload[x])
+        );
+      }
+    }
+    return payloadString;
+  }
+  
+  function _determineTransport(baseUrl, payloadString) {
+    return _transports[(payloadString.length >= (2048 - baseUrl.length - 1) ? str_xhr : str_image)];
```

Comment:
This brings two questions:
1. The backend we will have to support both POST and image logging, right?
2. Where does 2048 come from?